enclave/core/sgx/calls.c

```diff
@@ -775,6 +803,45 @@ oe_result_t oe_call_host_function_internal(
     /* Check the result */
     OE_CHECK(args->result);
 
+    return_args_ptr = (oe_call_function_return_args_t*)output_buffer;
+    /* Copy the marshaling struct from the host memory to avoid TOCTOU issues.
+     */
+    memcpy(
+        &return_args, return_args_ptr, sizeof(oe_call_function_return_args_t));
+
+    if (return_args.result == OE_OK)
+    {
+        /*
+         * Nonzero deepcopy_out_buffer and deepcopy_out_buffer_size fieldes
+         * indicate that there is deep-copied content that needs to be
+         * transmitted from the host. Also, ensuring that the content lies on
+         * host memory.
+         */
+        if (return_args.deepcopy_out_buffer &&
+            return_args.deepcopy_out_buffer_size &&
```

Comment:
In case there is a buffer, it would be better to error out if size == 0 or if the buffer does not lie out side the enclave.
Then we can be sure that downstream code won't encounter such a scenario.